Proposed 1.0

marijnh opened the pull request:
This is the state_object branch with the following changes:

 - The loose parser was also converted to object style
 - The plugin system is now based on method overriding rather than hooks, to make it easier to override various aspects of the system without us defining a hundred hooks. See the [jsx branch](https://github.com/marijnh/acorn/tree/jsx) for the JSX plugin ported over to this style.
 - The `acorn.tokenize` exported function was _removed_, and replaced with `acorn.tokenizer`, which simply returns a `Parser` instance that you can call `getToken` on to get the next token. (Such objects are now also iterable.)

@RReverser I want to bring something like this out as 1.0.0. Does it still sound okay to you?

marijnh commented:
Also, if you have any further things about the API you'd like to deprecate or change, now is a good time to start discussing/executing those, since I'd like to get our API in a state where we won't have to change it for a while.

Naddiseo commented:
As I mentioned in the other PR, perhaps consider making the prototype of `TokenType()` take an options parameter instead of four optional parameters so that reading a TokenType instantiation doesn't require looking at the implementation.

```javascript

var TokenType = exports.TokenType = function(label, opts) {
    this.label = label;
    this.keyword = opts.keyword;
    this.beforeExpr = !!opts.beforeExpr;
    this.isLoop = opts.kind === "loop";
    this.isAssign = opts.kind === "assign";
    this.prefix = opts.kind && opts.kind.indexOf("prefix") > -1;
    this.posfix = opts.kind && opts.kind.indexOf("postfix") > -1;
    this.binop = opts.binop || null;
    this.updateContext = null;
};


...
// Less readable
eq: new TokenType("=", true, null, "assign"),
// More readable
eq: new TokenType("=", {keyword: true, kind: "assign"})

```

RReverser commented:
+1 to @Naddiseo 

Also, let me bring agreed changes from ESTree (since most of the ES6-related issues are now resolved and agreed) before this release if you don't mind (I'll do it tomorrow or on weekend).

marijnh commented:
Patch 1255882 cleans up the `TokenType` constructor arguments. I guess I'm so used to Tern that I assume everyone can see argument names.

marijnh commented:
@RReverser Do you have a clear vision of how the splitting up and module building should work? In another project, I'm using ES6 modules, and have a pre-publish build step that runs babelify to combine the modules into a single library. But I don't have experience with bundling up a bunch of ES6 modules into _separate_ files (since we do not want to force everybody to load the loose parser and the tree walker when they don't need them). Do you have any ideas on how to approach that?

Also, would you be okay with moving the project to a semicolon-free code style? I've found omitting semicolons to make a surprisingly large (positive) difference in the look and feel of code.

RReverser commented:
> Do you have a clear vision of how the splitting up and module building should work? In another project, I'm using ES6 modules, and have a pre-publish build step that runs babelify to combine the modules into a single library. But I don't have experience with bundling up a bunch of ES6 modules into separate files (since we do not want to force everybody to load the loose parser and the tree walker when they don't need them). Do you have any ideas on how to approach that?

I think, for npm we can simply use ES6 modules with transpilation into CommonJS, so that people can manually load exactly what they need. As for the browser version, we can ship bundled regular parser + loose parser as separate files and for other cases (which are more specific and so rare) propose to do custom builds on their own as do [lodash](https://lodash.com/custom-builds), [jquery](https://github.com/jquery/jquery#how-to-build-your-own-jquery) and others.

> Also, would you be okay with moving the project to a semicolon-free code style? I've found omitting semicolons to make a surprisingly large (positive) difference in the look and feel of code.

Not to start bikeshedding, but personally, I find semicolon-free code style pretty unpleasant and would rather leave things as is.

marijnh commented:
Some incompatible changes in the options:

* Replaced `strictSemicolons` and `allowTrailingCommas` with `onInsertedSemicolon` and `onTrailingComma` (callbacks), since these are usually used for linting, and reporting on them can be done without aborting the whole parse. (The old behavior is trivial to emulate by passing a callback that raises an exception.)

* Renamed `forbidReserved` to `allowReserved`, inverted its meaning and default value.

RReverser commented:
I have a couple more things to be fixed on my side. Another thing we should consider (but it should not necessarily be in scope of 1.0) is lookahead - since tokenizer is now self-contained and not controlled by parser, it should be pretty cheap thing to do.

It would especially help to avoid complicated tricks for `{x, y}` vs `{x, y} = 1`, `[x, y]` vs `[x, y] = 1`, `async(x,y,z)` vs `async (x,y,z) => ...` and others.

Do you have any ideas on how we could implement such a "look-out" that would look-ahead after the corresponding closing bracket for current token if it's one of `{`/`[`/`(`?

Naddiseo commented:
@RReverser, that might be a useful utility function to have: skipNested(delim, (containedAST) => {}). If you have a '{' and want to skip to the matching '}', or get the ast in-between it'll register a callback to be called when the matching delim is found.

marijnh commented:
Is `{x} = {10}` valid ES6? I noticed Babel doesn't accept it, most likely because Acorn currently doesn't allow it.

Anyway, I'm not sure lookahead is a better solution that the current trick. It will definitely have a performance impact (reifying tokens, lots of them, since it has to happen for every `{}` or `[]` block) bigger than the current AST-node-transmuting trick.

RReverser commented:
I think it will be possible to optimize it later (like, caching corresponding closing brackets for nested stuff during the first loop and so), but avoiding tricks and enforcing spec-compliant parsing might give more important benefits.

RReverser commented:
@marijnh Looks like I'm done on my side, sorry for making it so long. So we just need to fix `yield`-related stuff and good to go.

RReverser commented:
Ok, looks like if you want to release now it's better to remove `SuperExpression` thing since new suggestions appeared and keeping it as `Identifier` would be safer. In any case, just let me know.

marijnh commented:
I haven't followed that discussion -- but if you think `SuperExpression` should go, do add a patch

RReverser commented:
It's not that it should go, but rather new arguments appeared when we almost agreed on whether it should be `SuperExpression` or just `Super`, and it's not clear yet which one will work. And I don't want to have release with controversial name, so if you want to release now, I'll just revert to `Identifier`.

marijnh commented:
I'm not sure I'll get around to releasing today -- I also want to set up the module/babel stuff first, and I want to give you a chance to properly look at those. So if you think the discussion will move forward in a few more days, let's wait it out (and put some pressure on the people involved to just make a decision already, since the issue seems somewhat trivial).

RReverser commented:
A lot of issues are pretty trivial, but you can't just stop bikeshedding :) Anyway, at least `Super` thing seems to evolve, so we should be fine in couple of days, if not - will release with `Identifier` and fix later.

marijnh commented:
@RReverser Re: omitting semicolons. Could you expand on why you find that unpleasant? Say we have... 

```javascript
export function parse(input, options) {
  let p = new Parser(options, input)
  let startPos = p.options.locations ? [p.pos, p.curPosition()] : p.pos
  p.nextToken()
  return p.parseTopLevel(p.options.program || p.startNodeAt(startPos))
}
```

Doesn't that look wonderful? It's almost as if we've left the language's C heritage behind.

marijnh commented:
(Context: I'm also working on a sizeable project completely written in ES6/semicolon-less style, and am finding it painful to move back to inserting semicolons everywhere. CodeMirror will be moved to that same style as well somewhere in the future.)

marijnh commented:
(@RReverser I re-applied your two patches from today on top of my work and force-pushed, since I had no idea how to cleanly rebase on top of them given the huge overhaul in file structure I applied.)

marijnh commented:
@sebmck  e88a543 should fix the for/in issue

RReverser commented:
Ok, I'll look into it. And I see that you went with semicolon-free style.

marijnh commented:
Don't look at 685b51e -- it breaks the build.

marijnh commented:
This is ready to review now.

I've also updated the `jsx` branch in a few minor ways to work with the new code.

RReverser commented:
Awesome work! Most likely I'll review in on Sunday and reply here.

marijnh commented:
A funny thing is that (on Node & Chrome) the module split actually made the code run faster. I'm have no explanation for that (except "no one can predict what the JIT will do"), but it's nice. The (output) code did get slightly bigger (~10% in lines) due to Babel output junk and extra wrapping closures.

RReverser commented:
> if (!node.body) node.body = []

@marijnh Any chance you would like to enforce braces for consistent styling? One-liners are sometimes cool, but not when you want to read code and have visual separation of concerns (condition and body).

marijnh commented:
> @marijnh Any chance you would agree to enforce braces for consistent styling?

I've never found the argument that braces make code easier to read very convincing. The difference between a condition and a body is perfectly unambiguous even without braces, and adding them for one liners feels like needless ceremony to me. So no, I really do not want to move to enforced-braces style.

RReverser commented:
> The difference between a condition and a body is perfectly unambiguous even without braces

IMO it depends on personality and habits, and having optional stuff like separate block for body included is a common denominator so that code is readable for everyone. Also, it's easier to read diffs in future when new statement to body is added (since only one new line appears instead of affecting entire line with both condition and previous statement).

marijnh commented:
> IMO it depends on personality and habits

It always does. But as the author of the project, I feel entitled to have the code fit my personality. I've never had trouble with this particular habit, and I think brevity is often a valuable property of code.

RReverser commented:
> But as the author of the project

Well, that's kinda valid argument but I want to continue contributing ES6, 7+ support to it, so trying to find some common denominator that would suit at least both of us.

marijnh commented:
Sure, but, in all seriousness, are brace-less bodies preventing you from effectively contributing to this project? I am willing to take concerns seriously (as I did with project attribution and modularization), but I just don't think that in this case, where we simply have different opinions on a very taste-based issue, I need to yield to your taste.

RReverser commented:
Well, I prefer KISS principle. Styling and tastes can be different, but rules how they are determines should be at least that simple that you don't even need to think of them while writing some new code.

For example, in general I like AirBnb style. They [point both options as valid](https://github.com/airbnb/javascript#blocks):

```javascript
// bad
if (test)
  return false;

// good
if (test) return false;

// good
if (test) {
  return false;
}
```

Those rules are pretty simple - if you have one statement and it nicely fits on one line, you put it there, if not - create a block. Just indentation is discarded as not visual separation.

But in current code I see things like i.e.:

```javascript
if (this.options.locations) { node.loc.end = pos[1]; pos = pos[0] }
...
if (startPos) {
    this.pos = startPos
    this.lineStart = Math.max(0, this.input.lastIndexOf("\n", startPos))
    this.curLine = this.input.slice(0, this.lineStart).split(lineBreak).length
}
...
if (!node.body) node.body = []
...
if (!topLevel)
  this.raise(this.start, "'import' and 'export' may only appear at the top level")
```

So it becomes unclear where border lies - sometimes we do put statements onto separate lines, sometimes we don't, sometimes we do one-liners only for one statement in body, sometimes it's one-liner even for multiple statements (block), sometimes for one statement we put body on the same line, sometimes on separate one and so on and so on.

What I'm asking here is not to change tastes but rather to have same consistent styling across all the code that can be easily followed and doesn't differ on case-by-case basis. Thoughts?

marijnh commented:
I like being free to lay out statements in any way that happens to feel appropriate in that context.

>    if (this.options.locations) { node.loc.end = pos[1]; pos = pos[0] }

This is a short aside, which happens to consist of more than a single statement.

>    if (startPos) {
>      this.pos = startPos
>      this.lineStart = Math.max(0, this.input.lastIndexOf("\n", startPos))
>      this.curLine = this.input.slice(0, this.lineStart).split(lineBreak).length
>    }

This is a significant branch doing multiple non-trivial things. A block can be appropriate for that.

>    if (!node.body) node.body = []

Another simple aside, but since it's a single statement, we can do away with the braces.

>    if (!topLevel)
>      this.raise(this.start, "'import' and 'export' may only appear at the top level")

Trivial, but rather long, so I gave the body its own line on that account.

Since the JS grammar does not require any deep analysis when distinguishing sub-statements from statements, I prefer to let the writer of the code decide how to format these. I don't expect all contributors to follow the same consistent style because _it really does not matter_. If you're used to code that enforces a certain narrower style, this might be slightly jarring, but it is something that's easy to get used to.

RReverser commented:
> I prefer to let the writer of the code decide how to format these

But you still did reformat my lines, so I though there should be some kind of rule that would be easy to follow.

But if we really don't care about coding style, I'm not sure there is something else for me to review given that functionality stays pretty much the same, just refactored. (Those commits that did change functionality look good to me and discussions were in relevant issues/PRs)

marijnh commented:
This has landed, and 1.0.0 has been published.
